flax/optim/radam.py

```diff
@@ -0,0 +1,111 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from .. import struct
+
+import jax.numpy as jnp
+from jax import lax
+
+import numpy as onp
+
+from .base import OptimizerDef
+
+
+@struct.dataclass
+class _RAdamHyperParams:
+  learning_rate: onp.ndarray
+  beta1: onp.ndarray
+  beta2: onp.ndarray
+  eps: onp.ndarray
+  weight_decay: onp.ndarray
+
+
+@struct.dataclass
+class _RAdamParamState:
+  grad_ema: onp.ndarray
+  grad_sq_ema: onp.ndarray
+
+
+class RAdam(OptimizerDef):
+  """RAdam optimizer."""
+
+  def __init__(self,
+               learning_rate=None,
+               beta1=0.9,
+               beta2=0.999,
+               eps=1e-8,
+               weight_decay=0.0):
+    """Constructor for the RAdam optimizer.
+
+    Args:
+      learning_rate: the step size used to update the parameters.
+      beta1: the coefficient used for the moving average of the
+        gradient (default: 0.9).
+      beta2: the coefficient used for the moving average of the
+        gradient magnitude (default: 0.999).
+      eps: the term added to the gradient magnitude estimate for
+        numerical stability.
+      weight_decay: AdamW style weight decay rate
+        (relative to learning rate).
+    """
+    hyper_params = _RAdamHyperParams(learning_rate, beta1, beta2, eps,
+                                     weight_decay)
+    super().__init__(hyper_params)
+
+  def init_param_state(self, param):
+    return _RAdamParamState(jnp.zeros_like(param), jnp.zeros_like(param))
+
+  def apply_param_gradient(self, step, hyper_params, param, state, grad):
+    assert hyper_params.learning_rate is not None, 'no learning rate provided.'
+    beta1 = hyper_params.beta1
+    beta2 = hyper_params.beta2
+    weight_decay = hyper_params.weight_decay
+    grad_sq = lax.square(grad)
+    grad_ema = beta1 * state.grad_ema + (1. - beta1) * grad
+    grad_sq_ema = beta2 * state.grad_sq_ema + (1. - beta2) * grad_sq
+
+    t = step + 1.
+    n_sma_inf = 2. / (1 - beta2) - 1.
+    n_sma_t = n_sma_inf - 2. * step * (beta2 ** t) / (1. - (beta2 ** t))
+
+    def high_variance(*args):
+      grad_sq_ema_ = jnp.sqrt(grad_sq_ema) + hyper_params.eps
+      step_size = jnp.sqrt((n_sma_t - 4) * (n_sma_t - 2) * n_sma_inf
+                           / ((n_sma_inf - 4) * (n_sma_inf - 2) *
+                              (n_sma_t))) * hyper_params.learning_rate \
+                           / (1 - (beta1 ** t))
+      new_param = param - ((grad_ema / grad_sq_ema_) *
+                           step_size * jnp.sqrt(1 - (beta2 ** t)))
+      return new_param
+
+    def low_variance(*args):
+      step_size = hyper_params.learning_rate / (1 - (beta1 ** t))
+      new_param = param - step_size
+      return new_param
+
+    new_param = lax.cond(n_sma_t >= 5,
```

Comment:
this will be verry slow on GPUs because cond will result in a sync between host and gpu for every parameter

Reply:
Also the constant here seems very arbitrary. Should it be configurable?

Reply:
It can be configurable, just followed the original author's implementation and since it was mentioned as a request in one of the optimizer feature request issues, I thought this would be a helpful contribution.